buildSrc/src/main/groovy/com/rocana/gradle/avro.groovy

```diff
@@ -0,0 +1,174 @@
+
+        // Create an Avro task for each config specified.
+        Task taskCompile = project.tasks.create("compile${container.name.capitalize()}Avro", AvroCompileTask.class) { task ->
+          task.source = sourceDirectory
+          task.generatedSourceDirectory = generatedSourceDirectory
+          task.fieldVisibility = container.fieldVisibility
+
+          if (container.templateDirectory != null) {
+            task.templateDirectory = container.templateDirectory.absolutePath
+          }
+        }
+
+        /*
+         * TODO: This will throw an exception if the user creates an Avro config
+         * for a source set that doesn't exist. We should throw a nicer error.
+         */
+
+        // Make the corresponding Java compilation task depend on the created Avro compile task.
+        project.tasks.getByName(compileTaskName).dependsOn(taskCompile)
+        project.sourceSets[container.name].java.srcDirs(taskCompile.generatedSourceDirectory)
+      }
+    }
+  }
+
+  def tasks = [AvroCompileTask.class]
+}
+
+class AvroContainer {
+
+  final String name
+
+  boolean createSetters
+  String fieldVisibility
+  String stringType
+  File templateDirectory
+  File sourceDirectory
+  File generatedSourceDirectory
+
+  AvroContainer(String name) {
+    this.name = name
+  }
+
+  @Override
+  String toString() {
+    return "{ name:$name, createSetters:$createSetters, fieldVisibility:$fieldVisibility, stringType:$stringType, templateDir:$templateDirectory, generatedSourceDirectory:$generatedSourceDirectory }"
+  }
+
+}
+
+class AvroCompileTask extends SourceTask {
+
+  @OutputDirectory
+  File generatedSourceDirectory
+
+  boolean generateSetters
+  String stringType
+  String fieldVisibility
+  String templateDirectory
+
+  AvroCompileTask() {
+    stringType = "String"
+    generateSetters = true
+
+    setDescription("Compile Avro schema files to Java source code")
+  }
+
+  @TaskAction
+  def compile() {
+    logger.info("Compiling {} files to {}", source.size(), generatedSourceDirectory)
+
+    Deque<File> toProcess = new ArrayDeque<>(source.getFiles())
+    Map<String, Schema> types = new HashMap<>()
+    Map<File, Schema> toCompile = new HashMap<>()
+
+    stringType = GenericData.StringType.valueOf(stringType)
+
+    while (!toProcess.isEmpty()) {
+      File file = toProcess.pop()
+
+      logger.info("Loading Avro schema file: {}. Known types: {}", file, types.keySet())
+
+      def parser = new Schema.Parser()
+      parser.addTypes(types)
+
+      Schema schema
+
+      try {
+        schema = parser.parse(file)
+
+        types.put(schema.name, schema)
+        toCompile.put(file, schema)
+      } catch (SchemaParseException e) {
+        if (e.getMessage().matches("unknown type")) {
+          logger.debug("Unknown type {} found - requeuing for parsing later.", e.getMessage())
+          toProcess.push(file)
```

Comment:
I think you're right and there's a bug here. There's also an infinite loop case if a schema file references a type that is not defined by any file that needs to be fixed.

Given the relative complexity of the outstanding PRs and their dependency, do you think it's ok to merge this as is and then immediately follow it with a bug fix? If not, I can just rebase the subsequent branches, but that may screw up the PRs since it will take a force push.